File: matmul/matmul.c
#include<stdio.h>
#include<time.h>
#include<stdlib.h>

#define DIMS 256
#define SIZE DIMS * sizeof(1.0)

// Currently its a square matrix
#define rowsA DIMS
#define colsA DIMS
#define rowsB DIMS
#define colsB DIMS

float A[SIZE][SIZE];
float B[SIZE][SIZE];
float C[SIZE][SIZE];

void matmul() {
    /*
    Arguably the simplest possible way to do matrix multiplication.
    There are much faster implementations out there.
    */
    size_t i, j, k;
    for (i = 0; i < rowsA; i++) {
        for (j = 0; j < colsB; j++) {
            for (k = 0; k < rowsB; k++) {
                C[i][j] += B[k][j] * A[j][k];
            }
        }
    }
}

float rand_float(float a) {
    /*
    returns a random float between 0.0 and a
    */
    float f = ((float)rand()/(float)(RAND_MAX)) * a;
    return f;
}

void random_fill() { 
    size_t i, j;
    for (i = 0; i < DIMS; i++){
        for (j = 0; j < DIMS; j++) {
            float temp_val = rand_float(10.0);
            A[i][j] = temp_val; 
        }
    }

    for (i = 0; i < DIMS; i++){
        for (j = 0; j < DIMS; j++) {
            float temp_val = rand_float(10.0);
            B[i][j] = temp_val; 
        }
    }
    
    for (i = 0; i < DIMS; i++){
        for (j = 0; j < DIMS; j++) {
            float temp_val = rand_float(10.0);
            C[i][j] = temp_val; 
        }
    }
}





int main() {
    printf("Multiplying a %u x %u matrix:\n",DIMS,DIMS);
    int i = 0;
    random_fill();
    clock_t t0, tf;
    double times[sizeof(1.0)*5];
    int iter = 5;
    while (i < iter){
        t0 = clock();
        matmul();
        tf = clock();
        double elap = (double)(tf - t0)/CLOCKS_PER_SEC;
        times[i] = elap;
        printf("Elapsed time = %.4lf s\n",elap);
        i++; 
    }

    double mean_time;
    for (i = 0; i < iter; i++) {
        mean_time += times[i];
    }
    mean_time = mean_time/iter;
    printf("Mean = %.4f s\n",mean_time);

    return 0;
}

File: matmul/matmul.py
#!/usr/bin/env python
import numpy as np
import time

def matmul(A,B) -> np.ndarray:
    rowsA = len(A[0])
    colsA = len(A)
    rowsB = len(B[0])
    colsB = len(B)

    C = np.zeros([colsA,rowsB])
    assert(colsA == rowsB),"Matrix dimensions are incompatible for matrix multiplication"
    for i in range(rowsA):
        for j in range(colsB):
            for k in range(rowsB):
                C[i][j] += B[i][k] * A[k][j]
    return C

def time_it(operation,args=None) -> float:
    t0 = time.monotonic();
    if args is not None:
        operation(*args)
    else:
        operation()
    tf = time.monotonic();
    elap = tf - t0
    return elap


def verify_matmul(A,B):
    my_res = matmul(A,B)
    np_res = np.matmul(A,B)
    for i in range(len(my_res)):
        for j in range(len(my_res[0])):    
            assert(my_res[i][j] == np_res[i][j]),f"Mismatch at {i,j}"

def main():
    DIMS = 128
    A = np.random.rand(DIMS,DIMS) 
    B = np.random.rand(DIMS,DIMS) 
    print(f"Multiplying a {DIMS} x {DIMS} Matrix")
    iter = 5
    i = 0
    times = []
    while i < iter:
        elap = time_it(matmul,args=[A,B]) 
        times.append(elap)
        print(f"Elapsed time = {elap:.4f} s")
        i += 1
    print(f"Mean = {np.mean(times):.4f} s")










if __name__ == "__main__":
    main()